Plugin/Magento/GiftCard/GenerateGiftCardAccountsOrderPlugin.php

```diff
@@ -0,0 +1,39 @@
+<?php
+/**
+ * Bolt magento2 plugin
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Bolt
+ * @package    Bolt_Boltpay
+ * @copyright  Copyright (c) 2019 Bolt Financial, Inc (https://www.bolt.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+namespace Bolt\Boltpay\Plugin\Magento\GiftCard;
+
+use Bolt\Boltpay\Model\Payment as BoltPayment;
+use Magento\Framework\Event\Observer;
+use Magento\Sales\Model\Order;
+
+class GenerateGiftCardAccountsOrderPlugin
+{
+    public function aroundExecute(
+        $subject,
+        callable $proceed,
+        Observer $observer
+    ) {
+        $event = $observer->getEvent();
+        /** @var Order $order */
+        $order =  $event->getOrder();
+        if ($order->getPayment()->getMethod() === BoltPayment::METHOD_CODE &&
+            in_array($order->getStatus(), [Order::STATE_PENDING_PAYMENT, Order::STATE_CANCELED])) {
```

Comment:
@danac-gs , as i see a class `\Magento\GiftCard\Observer\GenerateGiftCardAccountsOrder` call on event `sales_order_save_after`.
So, as i see the aroundPlugin functionality calling every time on order save. 

I guess we add here a status `Order::STATE_CLOSED` and `Order::STATE_HOLDED`, regarding that we already checking `Order::STATE_CANCELED`.

What do you think?

Reply:
@YevhenShBolt what we are trying to achieve here is to prevent creating gift cards when order record is created in pre-auth call, before passing the approval at Bolt (successful authorize call). Initially the order is created in pending_payment state, and we have two cases to cover: 
1. Failed payment - failed_payment hook triggers order deletion, no new state after pending_payment
2. Blacklisted cc - rejected_irreversible hook triggers order cancellation - new and final order state is canceled
Anything after pending_payment that is not canceled should not be a problem. After that point the order lifetime is similar as with other payment methods. We could be more strict but I don't think we need to. In the end we could allow gift card creation only if the order state is processing, but I did not want to alter the system more than necessary.